File: ApiBaseCore/ApiBaseCore.Domain/Models/Usuario.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace ApiBaseCore.Domain.Models;

/// <summary>
/// Define los roles que un usuario puede tener en el sistema.
/// </summary>
public enum RolUsuario
{
    User,
    Admin
}

/// <summary>
/// Representa la entidad principal de un usuario en la aplicación.
/// </summary>
public class Usuario
{
    #region Propiedades

    [Key]
    [JsonPropertyName("id")]
    public Guid Id { get; private set; }

    [Required]
    [EmailAddress]
    [MaxLength(100)]
    [JsonPropertyName("email")]
    public string Email { get; private set; }

    [Required]
    [MaxLength(100)]
    [JsonPropertyName("name")]
    public string Name { get; private set; }

    [Required]
    [MaxLength(20)]
    [JsonPropertyName("phone")]
    public string Phone { get; private set; }

    [JsonIgnore]
    public string? PasswordHash { get; private set; }

    [MaxLength(20)]
    [JsonPropertyName("tax_id")]
    public string? TaxId { get; private set; }

    [JsonPropertyName("created_at")]
    public DateTime CreatedAt { get; private set; }

    [Required]
    [JsonIgnore]
    public RolUsuario Rol { get; private set; }

    [JsonPropertyName("addresses")]
    public List<Address> Addresses { get; private set; } = new();

    // --- Technical/Auth Fields ---

    [JsonIgnore]
    public string? AvatarUrl { get; private set; }

    [JsonIgnore]
    public string? RefreshToken { get; private set; }

    [JsonIgnore]
    public DateTime? RefreshTokenExpiryTime { get; private set; }

    [Timestamp]
    [JsonIgnore]
    public byte[] RowVersion { get; private set; }

    [JsonIgnore]
    public string? EmailVerificationToken { get; private set; }

    [JsonIgnore]
    public bool IsEmailVerified { get; private set; }

    [JsonIgnore]
    public string? PasswordResetToken { get; private set; }

    [JsonIgnore]
    public DateTime? PasswordResetTokenExpiryTime { get; private set; }

    #endregion

    #region Constructores

    private Usuario()
    {
    }

    public Usuario(string name, string email, string phone, RolUsuario rol, string? taxId = null)
    {
        if (string.IsNullOrWhiteSpace(name))
            throw new ArgumentNullException(nameof(name), "El nombre es obligatorio.");
        if (string.IsNullOrWhiteSpace(email))
            throw new ArgumentNullException(nameof(email), "El email es obligatorio.");

        Id = Guid.NewGuid();
        Name = name;
        Email = email;
        Phone = phone;
        Rol = rol;
        TaxId = taxId;
        CreatedAt = DateTime.UtcNow;
        IsEmailVerified = false;
    }

    #endregion

    #region Métodos de Modificación

    public void EstablecerPasswordHash(string passwordHash)
    {
        if (string.IsNullOrWhiteSpace(passwordHash))
            throw new ArgumentNullException(nameof(passwordHash), "El hash de la contraseña no puede estar vacío.");
        PasswordHash = passwordHash;
    }

    public void UpdateProfile(string name, string phone, string? taxId)
    {
        if (!string.IsNullOrWhiteSpace(name)) Name = name;
        if (!string.IsNullOrWhiteSpace(phone)) Phone = phone;
        if (taxId != null) TaxId = taxId;
    }

    public void AddAddress(Address address)
    {
        Addresses.Add(address);
    }

    public void RemoveAddress(int addressId)
    {
        var address = Addresses.FirstOrDefault(a => a.Id == addressId);
        if (address != null)
        {
            Addresses.Remove(address);
        }
    }

    public void SetAvatarUrl(string nuevoUrl)
    {
        if (!string.IsNullOrWhiteSpace(nuevoUrl)) AvatarUrl = nuevoUrl;
    }
    
    public void SetRefreshToken(string refreshToken, DateTime expiryTime)
    {
        RefreshToken = refreshToken;
        RefreshTokenExpiryTime = expiryTime;
    }
    
    public void SetRol(RolUsuario rol)
    {
        Rol = rol;
    }

    public void SetEmailVerificationToken(string token)
    {
        EmailVerificationToken = token;
    }

    public void MarkEmailAsVerified()
    {
        IsEmailVerified = true;
        EmailVerificationToken = null;
    }

    public void SetPasswordResetToken(string token, DateTime expiryTime)
    {
        PasswordResetToken = token;
        PasswordResetTokenExpiryTime = expiryTime;
    }

    public void ClearPasswordResetToken()
    {
        PasswordResetToken = null;
        PasswordResetTokenExpiryTime = null;
    }

    #endregion
}

File: ApiBaseCoreJava/src/main/java/com/polancou/apibasecore/domain/models/Usuario.java
package com.polancou.apibasecore.domain.models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.polancou.apibasecore.domain.enums.RolUsuario;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "Usuarios", indexes = {
    @Index(name = "IX_Usuarios_TaxId", columnList = "TaxId", unique = true)
})
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Usuario {

    @Id
    @JsonProperty("id")
    private UUID id;

    @Column(nullable = false, length = 100)
    @JsonProperty("email")
    private String email;

    @Column(nullable = false, length = 100)
    @JsonProperty("name")
    private String name;

    @Column(nullable = false, length = 20)
    @JsonProperty("phone")
    private String phone;

    @JsonIgnore
    @Column(name = "PasswordHash")
    private String passwordHash;

    @Column(length = 20, unique = true)
    @JsonProperty("tax_id")
    private String taxId;

    @JsonProperty("created_at")
    private LocalDateTime createdAt;

    @Enumerated(EnumType.ORDINAL) // Matching C# default (int)
    @Column(nullable = false)
    @JsonIgnore // In C# it was JsonIgnore, but usually we want role in response? C# code had [JsonIgnore]
    private RolUsuario rol;

    @OneToMany(mappedBy = "usuario", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonProperty("addresses")
    private List<Address> addresses = new ArrayList<>();

    // --- Technical/Auth Fields ---

    @JsonIgnore
    private String avatarUrl;

    @JsonIgnore
    private String refreshToken;

    @JsonIgnore
    private LocalDateTime refreshTokenExpiryTime;

    @Version
    @JsonIgnore
    private Long rowVersion; // Hibernate uses Long or Integer for versioning, check compatibility with byte[]

    @JsonIgnore
    private String emailVerificationToken;

    @JsonIgnore
    private boolean isEmailVerified;

    @JsonIgnore
    private String passwordResetToken;

    @JsonIgnore
    private LocalDateTime passwordResetTokenExpiryTime;

    // Constructors

    public Usuario(String name, String email, String phone, RolUsuario rol, String taxId) {
        if (name == null || name.isBlank()) throw new IllegalArgumentException("El nombre es obligatorio.");
        if (email == null || email.isBlank()) throw new IllegalArgumentException("El email es obligatorio.");

        this.id = UUID.randomUUID();
        this.name = name;
        this.email = email;
        this.phone = phone;
        this.rol = rol;
        this.taxId = taxId;
        this.createdAt = LocalDateTime.now(ZoneId.of("UTC")); // Similar to DateTime.UtcNow
        this.isEmailVerified = false;
    }

    // Methods
    public void setId(UUID id) {
        this.id = id;
    }

    public void establecerPasswordHash(String passwordHash) {
        if (passwordHash == null || passwordHash.isBlank())
            throw new IllegalArgumentException("El hash de la contraseña no puede estar vacío.");
        this.passwordHash = passwordHash;
    }

    public void updateProfile(String name, String phone, String taxId) {
        if (name != null && !name.isBlank()) this.name = name;
        if (phone != null && !phone.isBlank()) this.phone = phone;
        if (taxId != null) this.taxId = taxId;
    }

    public void addAddress(Address address) {
        addresses.add(address);
        address.setUsuario(this);
    }

    public void removeAddress(Integer addressId) {
        addresses.removeIf(a -> a.getId().equals(addressId));
    }

    public void setAvatarUrl(String nuevoUrl) {
        if (nuevoUrl != null && !nuevoUrl.isBlank()) this.avatarUrl = nuevoUrl;
    }

    public void setRefreshToken(String refreshToken, LocalDateTime expiryTime) {
        this.refreshToken = refreshToken;
        this.refreshTokenExpiryTime = expiryTime;
    }

    public void setRol(RolUsuario rol) {
        this.rol = rol;
    }

    public void setEmailVerificationToken(String token) {
        this.emailVerificationToken = token;
    }

    public void markEmailAsVerified() {
        this.isEmailVerified = true;
        this.emailVerificationToken = null;
    }

    public void setPasswordResetToken(String token, LocalDateTime expiryTime) {
        this.passwordResetToken = token;
        this.passwordResetTokenExpiryTime = expiryTime;
    }

    public void clearPasswordResetToken() {
        this.passwordResetToken = null;
        this.passwordResetTokenExpiryTime = null;
    }
}
